lib/Dialect/Triton/IR/Ops.cpp

```diff
@@ -546,19 +546,24 @@ LogicalResult ReduceOp::inferReturnTypes(
   return success();
 }
 
-LogicalResult ReduceOp::verify() {
-  if (this->getOperands().size() < 1) {
-    return this->emitOpError() << "must have at least 1 operand";
+// Helpers for Reductions and Scans
+template <class Op> LogicalResult verifyReduceScan(Op &op) {
+  if (op.getOperands().empty()) {
+    return op.emitOpError() << "must have at least 1 operand";
+  }
+  if (op.getNumOperands() != op.getNumResults()) {
+    return op.emitOpError() << "must have the same number of inputs as outputs";
   }
-  for (const auto &operand : this->getOperands()) {
-    if (!dyn_cast<RankedTensorType>(operand.getType())) {
-      return this->emitOpError() << "operands must be RankedTensorType";
+
+  for (auto [opTy, resTy] :
+       llvm::zip(op.getOperandTypes(), op.getResultTypes())) {
+    if (opTy != resTy) {
+      return op.emitOpError() << "operand types and result types must agree";
```

Comment:
looking at the code again I wonder if we need that. Why don't we get an error if `inferReturnTypes` has a different type than what is expected?

Reply:
This is effectively doing exactly that. `getResultTypes` will be given by `inferReturnTypes`.

Reply:
ah I see, yes makes sense